File: data_processing/step4_sumstats.R
.libPaths("/home/ivm/R/x86_64-pc-linux-gnu-library/4.4")

#install.packages("readr")
#install.packages("dplyr")
#install.packages("optparse")
#install.packages("e1071")
#install.packages("lubridate")
library(plyr)
library(readr)
library(dplyr)
library(optparse)
library(e1071)
library(lubridate)
library(arrow)

options(width=250)

args_list <- list(
  make_option(c("-r", "--res_dir"), action="store", type="character", default="/home/ivm/valid/data/processed_data/step5_data/1_year_buffer/",
              help="Path to results directory."),
  make_option(c("-f", "--file_path"), action="store", type="character", default="/home/ivm/valid/data/processed_data/step4_labels/",
              help="Path to data. Needs to contain both data and metadata (same name with _name.csv) at the end"),
  make_option(c("-d", "--file_name"), action="store", type="character", default="2025-01-31",help="Date of original processed file."),
  make_option(c("--file_path_labels"), action="store", type="character", default="",help="If label file different. [default: '']"),
  make_option(c("--custom_file_name"), action="store", type="character", default="",help="If label file different. [default: '']")
  
)
parser <- OptionParser(option_list=args_list)
args <- parse_args(parser, positional_arguments=0)$options
print(args)

date = Sys.Date()
file_path_data <- paste0(args$file_path, args$file_name, ".parquet")

data <- tibble::as_tibble(arrow::read_parquet(file_path_data))
if(args$file_path_labels != "") {
  labels <-  tibble::as_tibble(arrow::read_parquet(args$file_path_labels))
  out_file_path <- paste0(args$res_dir, args$custom_file_name, "_sumstats.parquet")
} else {
  if(args$custom_file_name == ""){
    out_file_path <- paste0(args$res_dir, args$file_name, "_sumstats.parquet")
  } else {
    out_file_path <- paste0(args$res_dir, args$custom_file_name, "_sumstats.parquet")
  }
  labels <- tibble::as_tibble(arrow::read_parquet(paste0(args$file_path, args$file_name, "_labels.parquet")))
}

data <- left_join(labels %>% select(FINNGENID, START_DATE, SET), data) %>% arrange(FINNGENID) 
print(data)

mean_val <- mean(data %>% dplyr::filter(SET == 0) %>% pull(VALUE))
set.seed(10234)

sumstats <- data  %>% group_by(FINNGENID) %>% dplyr::filter(DATE < START_DATE, !is.na(VALUE)) %>% dplyr::arrange(desc(DATE)) %>%
  dplyr::reframe(MIN=min(VALUE), 
                 MAX=max(VALUE), 
                 SD=sd(VALUE), 
                 MEAN=mean(VALUE), 
                 MEDIAN=median(VALUE), 
                 SUM=sum(VALUE), 
                 KURT=e1071::kurtosis(VALUE), 
                 SKEW=e1071::skewness(VALUE), 
                 ABS_ENERG=sum(VALUE**2),
                 SUM_ABS_CHANGE=sum(abs(VALUE-lead(VALUE)), na.rm=TRUE),
                 MEAN_ABS_CHANGE=mean(abs(VALUE-lead(VALUE)), na.rm=TRUE),
                 MAX_ABS_CHANGE=max(abs(lead(VALUE)-VALUE), na.rm=TRUE),
                 MAX_CHANGE=max(lead(VALUE)-VALUE, na.rm=TRUE),
                 MEAN_CHANGE=mean(VALUE-lag(VALUE), na.rm=TRUE),
                 UNIQUE_VALS=length(unique(round(VALUE))),
                 SEQ_LEN=length(VALUE),
                 QUANT_25=quantile(VALUE, probs=c(0.25)),
                 QUANT_75=quantile(VALUE, probs=c(0.75)),
                 IDX_QUANT_100=VALUE[quantile(1:n(), probs=c(0))],
                 IDX_QUANT_50=VALUE[quantile(1:n(), probs=c(0.5))],
                 IDX_QUANT_0=VALUE[quantile(1:n(), probs=c(1))],
                 ABNORM=sum(ABNORM_CUSTOM),
                 MIN_LOC=lubridate::time_length(DATE[which.min(VALUE)]%--%START_DATE, "days"), 
                 MAX_LOC=lubridate::time_length(DATE[which.max(VALUE)]%--%START_DATE, "days"),
                 FIRST_LAST=lubridate::time_length(min(DATE)%--%max(DATE), "days"),
                 LAST_VAL_DATE=max(DATE),
                 SET=SET) %>% distinct() %>% ungroup() 
sumstats %>%mutate(FIRST_LAST=round(FIRST_LAST/365.25)) %>%pull(FIRST_LAST) %>% round() %>% table()
sumstats <- dplyr::mutate(sumstats, MAX_CHANGE=ifelse(is.infinite(MAX_CHANGE), NA, MAX_CHANGE), MAX_ABS_CHANGE=ifelse(is.infinite(MAX_ABS_CHANGE), NA, MAX_ABS_CHANGE))
sumstats <- dplyr::select(sumstats, -UNIQUE_VALS)
sumstats$LAST_VAL_DATE <- as.Date(sumstats$LAST_VAL_DATE)

## Adding sumstats for missing values
# Missing data-imputation
if(args$file_path_labels == "")  {
  missing_data <- dplyr::group_by(data, FINNGENID) %>% dplyr::filter(n()==1, is.na(VALUE)) %>% ungroup()
  train_sumstats <- dplyr::filter(sumstats,SET==0)
  
  missing_data_sumstats <- missing_data %>% dplyr::mutate(MIN=mean(train_sumstats$MIN, na.rm=TRUE), 
                                                          MAX=mean(train_sumstats$MAX, na.rm=TRUE), 
                                                          SD=NA, 
                                                          MEAN=mean(train_sumstats$MEAN, na.rm=TRUE), 
                                                          MEDIAN=mean(train_sumstats$MEDIAN, na.rm=TRUE),
                                                          SUM=mean(train_sumstats$SUM, na.rm=TRUE), 
                                                          KURT=NA, 
                                                          SKEW=NA, 
                                                          ABS_ENERG=mean(train_sumstats$ABS_ENERG, na.rm=TRUE), 
                                                          SUM_ABS_CHANGE=0,
                                                          MEAN_ABS_CHANGE=NA,
                                                          MAX_CHANGE=NA,
                                                          MAX_ABS_CHANGE=NA,
                                                          MEAN_CHANGE=NA, 
                                                          QUANT_25=mean(train_sumstats$QUANT_25, na.rm=TRUE), 
                                                          QUANT_75=mean(train_sumstats$QUANT_75, na.rm=TRUE),
                                                          IDX_QUANT_0=mean(train_sumstats$IDX_QUANT_0, na.rm=TRUE), 
                                                          IDX_QUANT_50=mean(train_sumstats$IDX_QUANT_50, na.rm=TRUE),
                                                          IDX_QUANT_100=mean(train_sumstats$IDX_QUANT_100, na.rm=TRUE), 
                                                          MIN_LOC=mean(train_sumstats$MIN_LOC, na.rm=TRUE),
                                                          MAX_LOC=mean(train_sumstats$MAX_LOC, na.rm=TRUE),
                                                          ABNORM=0,
                                                          SEQ_LEN=0,
                                                          FIRST_LAST=0.0,
                                                          LAST_VAL_DATE=NA)
  missing_data_sumstats <- missing_data_sumstats[colnames(sumstats)]
  print(missing_data_sumstats)
  sumstats <- rbind(sumstats, missing_data_sumstats)
}

dir.create(args$res_dir, showWarnings=FALSE)
arrow::write_parquet(sumstats, out_file_path)

File: data_processing/step4_sumstats.py


# Custom utils
import sys

sys.path.append(("/home/ivm/valid/scripts/utils/"))
from general_utils import get_date, init_logging, Timer, read_file
from processing_utils import egfr_ckdepi2021_transform

# Standard stuff
import polars as pl
import argparse
# Logging
import logging
logger = logging.getLogger(__name__)
from datetime import datetime



def get_parser_arguments():
    #### Parsing and logging
    parser = argparse.ArgumentParser()
    # Saving info
    parser.add_argument("--res_dir", type=str, help="Path to the results directory", required=True)
    parser.add_argument("--file_path", type=str, help="Path to the data file", required=True)
    parser.add_argument("--file_name_start", type=str, help="Name of the data file", required=True)
    parser.add_argument("--file_path_labels", type=str, help="Path to the diagnosis data file", default="")
    parser.add_argument("--file_path_data", type=str, help="Path to the diagnosis data file", default="")
    parser.add_argument("--file_path_data_2", type=str, help="Path to the diagnosis data file", default="")
    parser.add_argument("--lab_name", type=str, help="Readable name of the measurement value for file naming.", required=True)
    parser.add_argument("--start_date", type=str, default="", help="Date to filter before")
    parser.add_argument("--mean_impute", type=int, default=1, help="Whether to impute mean for those with missing. [Options: 0 (No), 1 (Yes some), 2 (Yes all)]")
    parser.add_argument("--interpolate", type=int, default=0, help="Whether to first mean aggregate the data for each month and then interpolate the missing months.")

    # Settings
    args = parser.parse_args()
    return(args)


#do the main file functions and runs 
if __name__ == "__main__":
    # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # 
    #                 Initial setup                                           #
    # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #   
    timer = Timer()
    args = get_parser_arguments()
    init_logging(args.res_dir, args.lab_name, logger, args)
    if args.mean_impute >= 1 and not args.interpolate >= 1:
        out_file_name = args.file_name_start+"_sumstats_"
    elif args.mean_impute >= 1 and args.interpolate >= 1:
        out_file_name = args.file_name_start+"_sumstats_int_"
    elif args.mean_impute == 0 and args.interpolate >= 1:
        out_file_name = args.file_name_start+"_sumstats_ni_int_"
    else:
        out_file_name = args.file_name_start+"_sumstats_ni_"
    if args.start_date != "": 
        out_file_name = out_file_name + "pred" + str(int(args.start_date[0:4])+1) + "_"
    out_file_name = out_file_name + get_date()
    
    # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # 
    #                 Get data                                                #
    # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #   
    
    
    if args.file_path_labels != "":
        print(args.file_path_data)
        data = read_file(args.file_path_data)
        if args.file_path_data_2 != "":
            data_2 = read_file(args.file_path_data_2)
            data = pl.concat([data, data_2])
        labels = read_file(args.file_path_labels)
    else:
        data = read_file(args.file_path+args.file_name_start+".parquet")
        labels = read_file(args.file_path+args.file_name_start+"_labels.parquet")
    data = data.join(labels, on="FINNGENID", how="right", coalesce=True)
    if args.start_date != "": 
        try:
            data = data.with_columns(pl.Series("START_DATE", [datetime.strptime(args.start_date, "%Y-%m-%d")]*data.height))
        except:
            data = data.with_columns(pl.Series("START_DATE", [datetime.strptime(args.start_date, "%Y-%m-%d %H:%M:%S")]*data.height))

    if args.interpolate == 1:
        print(data.columns)
        no_data = data.filter(pl.col.VALUE.is_null()).select(["FINNGENID", "SEX", "EVENT_AGE", "VALUE", "DATE", "START_DATE", "SET", "ABNORM_CUSTOM"])
        data = data.filter(pl.col.VALUE.is_not_null())
        month_data = (data
                      .filter(pl.col.DATE<pl.col.START_DATE)
                      .with_columns(pl.col.DATE.dt.truncate("1mo").alias("DATE"))
                      .group_by("FINNGENID", "DATE")
                      .agg(pl.col.VALUE.mean(),
                           pl.col.SEX.first(),
                           pl.col.START_DATE.first(),
                           pl.col.SET.first(),
                           pl.col.EVENT_AGE.min()+((pl.col.EVENT_AGE.max()-pl.col.EVENT_AGE.min())/2))
        )
        all_months = (month_data
                         .group_by("FINNGENID")
                         .agg(pl.date_range(pl.col.DATE.min(), pl.col.DATE.max(), interval="1mo").alias("DATE"))
                         .explode("DATE")
        )
        data = (month_data
                    .join(all_months, on=["FINNGENID", "DATE"], how="right")
                    .sort(["FINNGENID", "DATE"])
                    .with_columns(pl.col.VALUE.interpolate().over("FINNGENID").alias("VALUE"),
                                  pl.col.EVENT_AGE.interpolate().over("FINNGENID").alias("EVENT_AGE"),
                                  pl.col.SEX.first().over("FINNGENID").alias("SEX")
                                 )
                    
        ).select(["FINNGENID", "SEX", "EVENT_AGE", "VALUE", "DATE", "START_DATE", "SET"])
        print(data)
        data = egfr_ckdepi2021_transform(data, "VALUE")
        data = data.with_columns(pl.when(pl.col.VALUE<60).then(pl.lit(1)).otherwise(pl.lit(0)).alias("ABNORM_CUSTOM").cast(pl.Float64))
        data = pl.concat([no_data, data])
        print(data)

    # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # 
    #                 Summary statistics                                      #
    # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #   
    sumstats = (data
                .filter(pl.col.DATE<pl.col.START_DATE)
                .filter((pl.col.VALUE.is_not_null()))
                .sort(["FINNGENID", "DATE"], descending=False)
                .group_by(["FINNGENID", "SET"])
                .agg(
                pl.col.VALUE.mean().alias("MEAN"),
                pl.lit(0).alias("NO_HISTORY"),
                pl.col.VALUE.std().alias("SD"),
                pl.col.VALUE.min().alias("MIN"),
                pl.col.VALUE.max().alias("MAX"),
                pl.col.VALUE.sum().alias("SUM"),
                pl.col.VALUE.kurtosis().alias("KURT"),
                pl.col.VALUE.skew().alias("SKEW"),
                (pl.col.VALUE**2).sum().alias("ABS_ENERG"),
                (pl.col.VALUE-pl.col.VALUE.shift(1)).sum().alias("SUM_CHANGE"),
                (pl.col.VALUE-pl.col.VALUE.shift(1)).mean().alias("MEAN_CHANGE"),
                (pl.col.VALUE.count()).cast(pl.Int64).alias("SEQ_LEN"),
                (pl.col.VALUE.quantile(0.25)).alias("QUANT_25"),
                (pl.col.VALUE.quantile(0.75)).alias("QUANT_75"),
                (pl.col.VALUE.get(pl.col.DATE.arg_min())).alias("IDX_QUANT_0"),
                (pl.col.VALUE.get(pl.col.DATE.arg_max())).alias("IDX_QUANT_100"),
                (pl.col.ABNORM_CUSTOM.sum()).cast(pl.Int64).alias("ABNORM"),
                # Distance of minimum compared to prediction in days
                (((pl.col.DATE.get(pl.col.VALUE.arg_min())-pl.col.START_DATE.first()).dt.total_days()/365.25).abs()).alias("MIN_LOC"),
                (((pl.col.DATE.get(pl.col.VALUE.arg_max())-pl.col.START_DATE.first()).dt.total_days()/365.25).abs()).alias("MAX_LOC"),
                ((pl.col.DATE.max()-pl.col.DATE.min()).dt.total_days()/365.25).alias("FIRST_LAST"),
                (pl.col.DATE.max()).alias("LAST_VAL_DATE")
                )
    )
    # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # 
    #                 Mean for those with missing data                        #
    # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #  
    sumstats_train = sumstats.filter(pl.col.SET==0)
    if args.file_path_labels=="" and args.mean_impute >= 1:
        # atm this only works for the original data as labels then has empty rows. Otherwise we just dont have data.
        missing_data = data.filter((pl.col.VALUE.is_null())&((pl.len()==1).over("FINNGENID"))).select("FINNGENID", "SET")

        if args.mean_impute == 1:
            missing_data = (missing_data.with_columns(
                                pl.Series("MIN", [sumstats_train["MIN"].mean()]*missing_data.height),
                                pl.Series("MAX", [sumstats_train["MAX"].mean()]*missing_data.height),
                                pl.Series("MEAN", [sumstats_train["MEAN"].mean()]*missing_data.height),
                                pl.Series("NO_HISTORY", [1]*missing_data.height).cast(pl.Int32, strict=False),
                                pl.Series("SUM", [sumstats_train["SUM"].mean()]*missing_data.height),
                                pl.Series("ABS_ENERG", [sumstats_train["ABS_ENERG"].mean()]*missing_data.height),
                                pl.Series("QUANT_25", [sumstats_train["QUANT_25"].mean()]*missing_data.height),
                                pl.Series("QUANT_75", [sumstats_train["QUANT_75"].mean()]*missing_data.height),
                                pl.Series("IDX_QUANT_0", [sumstats_train["IDX_QUANT_0"].mean()]*missing_data.height),
                                pl.Series("IDX_QUANT_100", [sumstats_train["IDX_QUANT_100"].mean()]*missing_data.height),
                                pl.Series("MIN_LOC", [sumstats_train["MIN_LOC"].mean()]*missing_data.height).cast(pl.Float64),
                                pl.Series("MAX_LOC", [sumstats_train["MAX_LOC"].mean()]*missing_data.height).cast(pl.Float64),
                                pl.Series("KURT", [None]*missing_data.height).cast(pl.Float64, strict=False),
                                pl.Series("SKEW", [None]*missing_data.height).cast(pl.Float64, strict=False),
                                pl.Series("SD", [None]*missing_data.height).cast(pl.Float64, strict=False),
                                pl.Series("MEAN_CHANGE", [None]*missing_data.height).cast(pl.Float64, strict=False),
                                pl.Series("SUM_CHANGE", [0]*missing_data.height).cast(pl.Float64, strict=False),
                                pl.Series("LAST_VAL_DATE", [None]*missing_data.height).cast(pl.Date, strict=False),
                                pl.Series("SEQ_LEN", [0]*missing_data.height).cast(pl.Int64, strict=False),
                                pl.Series("FIRST_LAST", [0]*missing_data.height).cast(pl.Float64, strict=False),
                                pl.Series("ABNORM", [0]*missing_data.height).cast(pl.Int64, strict=False),
                            )
            )
        elif args.mean_impute == 2:
           missing_data = (missing_data.with_columns(
                                pl.Series("MIN", [sumstats_train["MIN"].mean()]*missing_data.height),
                                pl.Series("MAX", [sumstats_train["MAX"].mean()]*missing_data.height),
                                pl.Series("MEAN", [sumstats_train["MEAN"].mean()]*missing_data.height),
                                pl.Series("NO_HISTORY", [1]*missing_data.height).cast(pl.Int32, strict=False),
                                pl.Series("SUM", [sumstats_train["SUM"].mean()]*missing_data.height),
                                pl.Series("ABS_ENERG", [sumstats_train["ABS_ENERG"].mean()]*missing_data.height),
                                pl.Series("QUANT_25", [sumstats_train["QUANT_25"].mean()]*missing_data.height),
                                pl.Series("QUANT_75", [sumstats_train["QUANT_75"].mean()]*missing_data.height),
                                pl.Series("IDX_QUANT_0", [sumstats_train["IDX_QUANT_0"].mean()]*missing_data.height),
                                pl.Series("IDX_QUANT_100", [sumstats_train["IDX_QUANT_100"].mean()]*missing_data.height),
                                pl.Series("MIN_LOC", [sumstats_train["MIN_LOC"].mean()]*missing_data.height).cast(pl.Float64),
                                pl.Series("MAX_LOC", [sumstats_train["MAX_LOC"].mean()]*missing_data.height).cast(pl.Float64),
                                pl.Series("KURT", [sumstats_train["KURT"].mean()]*missing_data.height).cast(pl.Float64, strict=False),
                                pl.Series("SKEW", [sumstats_train["SKEW"].mean()]*missing_data.height).cast(pl.Float64, strict=False),
                                pl.Series("SD", [sumstats_train["SD"].mean()]*missing_data.height).cast(pl.Float64, strict=False),
                                pl.Series("MEAN_CHANGE", [sumstats_train["MEAN_CHANGE"].mean()]*missing_data.height).cast(pl.Float64, strict=False),
                                pl.Series("SUM_CHANGE", [0]*missing_data.height).cast(pl.Float64, strict=False),
                                pl.Series("LAST_VAL_DATE", [datetime(2013,1,1)]*missing_data.height).cast(pl.Date, strict=False),
                                pl.Series("SEQ_LEN", [0]*missing_data.height).cast(pl.Int64, strict=False),
                                pl.Series("FIRST_LAST", [0]*missing_data.height).cast(pl.Float64, strict=False),
                                pl.Series("ABNORM", [0]*missing_data.height).cast(pl.Int64, strict=False),
                            )
            )
            
            
        print(missing_data)
        sumstats = pl.concat([sumstats, missing_data.select(sumstats.columns)])
    print(sumstats)
    # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # 
    #                 Saving                                                  #
    # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #  
    sumstats.write_parquet(args.res_dir+out_file_name+".parquet")